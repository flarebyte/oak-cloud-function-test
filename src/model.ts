export interface OakBase {
  name: string;
  description: string;
  flags: string[];
}

/**
 * Metadata about a status that a function or cloud service would return
 */
export type OakStatus = OakBase;

/**
 * Metadata about a cloud service
 * @example aws:s3
 */
export type OakService = OakBase;

/**
 * Metadata about an operation available on a cloud service
 */
export interface OakServiceOperation extends OakBase {
  functionName: string;
  service: OakService;
  version: string;
  systemFlagsDict: {
    [name: string]: string;
  };
}

export type Anything = string | number | boolean | object;

export type OakResource = OakBase;

/**
 * Configuration of a service
 * @example read-from-s3
 */
export interface OakBusinessOperation extends OakBase {
  resource: OakResource;
}

export interface OakAction extends OakBase {
  functionName: string;
  version: string;
  customStatusDict: {
    [name: string]: OakStatus;
  };
  systemFlagsDict: {
    [name: string]: string;
  };
}

export interface OakFunction extends OakBase {
  functionName: string;
  version: string;
  customStatusDict: {
    [name: string]: OakStatus;
  };

  systemFlagsDict: {
    [name: string]: string;
  };
}
/**
 * An event that represent a payload
 */
export interface OakEvent {
  comment: string;
  payload: object;
  flags: string[];
}

export interface OakRequestEvent extends OakEvent {
  businessOperation: OakBusinessOperation;
  caller: string;
  serviceParams: object;
  systemFlags: string[];
}

/**
 * An event representing a response from a service
 */
export interface OakResponseEvent extends OakEvent {
  status: OakStatus;
}

export interface OakActionRequestEvent extends OakEvent {
  caller: string;
  params: object;
  systemFlags: string[];
}

export interface OakEventTransaction {
  id: number;
  serviceOperation: OakServiceOperation;
  request: OakRequestEvent;
  response: OakResponseEvent;
  nanoSeconds: string;
}

export interface OakActionEventTransaction {
  id: number;
  action: OakAction;
  request: OakActionRequestEvent;
  response: OakResponseEvent;
}

export interface OakActionAndRequest {
  action: OakAction;
  request: OakActionRequestEvent;
}

/**
 * Metadata about a service and the available operations
 */
export interface OakServiceData extends OakBase {
  service: OakService;
  version: string;
  serviceOpDict: {
    [name: string]: OakServiceOperation;
  };
  customStatusDict: {
    [name: string]: OakStatus;
  };
}
/**
 * A context object for the engine
 */
export interface OakEngineContext {
  transactions: OakEventTransaction[];
  actionTransactions: OakActionEventTransaction[];
  systemFlags: string[];
  businessOperationFlags: {
    [name: string]: string[];
  };
  actionFlags: {
    [name: string]: string[];
  };
  functionFlags: {
    [name: string]: string[];
  };
}

/**
 * A low level call to the cloud service that should be simulated 
 */
export type OakCall = (
  context: OakEngineContext,
  request: OakRequestEvent
) => Promise<OakResponseEvent>;

export type OakCtxCall = (
  request: OakRequestEvent
) => Promise<OakResponseEvent>;

export interface OakBeforeCallInterceptor {
  call: OakCall;
  pass: boolean;
}
export type OakBeforeCall = (
  serviceOperation: OakServiceOperation,
  context: OakEngineContext,
  request: OakRequestEvent
) => OakBeforeCallInterceptor;

export interface OakActionCompanion {
  explicitCall: {
    [name: string]: OakCall;
  };
  call: {
    [name: string]: OakCtxCall;
  };
  callServiceOperationDict: {
    [name: string]: OakServiceOperation;
  };
}

export type OakActionCall = (
  context: OakEngineContext,
  companion: OakActionCompanion,
  request: OakActionRequestEvent
) => Promise<OakResponseEvent>;

export type OakActionCtxCall = (
  request: OakActionRequestEvent
) => Promise<OakResponseEvent>;

export interface OakFunctionCompanion {
  explicitCall: {
    [name: string]: OakActionCall;
  };
  call: {
    [name: string]: OakActionCtxCall;
  };
  actionDict: {
    [name: string]: OakAction;
  };
}

export type OakCallWrapper = (
  serviceOperation: OakServiceOperation,
  call: OakCall
) => OakCtxCall;

export type OakActionCallWrapper = (
  action: OakAction,
  call: OakActionCall
) => OakActionCtxCall;

export interface OakServiceOpToCall {
  so: OakServiceOperation;
  call: OakCall;
}

export interface OakActionToCall {
  action: OakAction;
  call: OakActionCall;
}
