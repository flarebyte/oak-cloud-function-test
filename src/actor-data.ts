import { OakAction } from './model';
import { version } from './version';
import { OakFunction } from '../src/model';

const template = {
  version,
  flags: [],
  customStatusDict: {},
  systemFlagsDict: {},
};

export const beforeEachAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/before-each',
  functionName: 'beforeEachAction',
  description: 'Runs a function before each of the tests is run',
};

export const beforeTestAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/before-test',
  functionName: 'beforeTestAction',
  description: 'Runs a function before this test is run',
};

export const testAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/test',
  functionName: 'testAction',
  description: 'Runs as part of the core test',
};

export const beforeEachFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/before-each',
  functionName: 'beforeEachFunc',
  description: 'Runs a function before each of the tests is run',
};

export const beforeTestFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/before-test',
  functionName: 'beforeTestFunc',
  description: 'Runs a function before this test is run',
};

export const testFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/test',
  functionName: 'testFunc',
  description: 'Runs as part of the core test',
};

export const actorData = {
  actionDict: {
    beforeEach: beforeEachAction,
    beforeTest: beforeTestAction,
    test: testAction,
  },

  funcDict: {
    beforeEach: beforeEachFunc,
    beforeTest: beforeTestFunc,
    test: testFunc,
  },
};
