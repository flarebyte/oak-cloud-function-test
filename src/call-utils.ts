const senderSeparator = ' -> ';
const noEmptyString = (value: string): boolean => value.length > 0;

export const joinSenders = (senders: string[]): string =>
  senders.filter(noEmptyString).join(senderSeparator);
export const splitSenders = (sender: string): string[] =>
  sender.split(senderSeparator).filter(noEmptyString);

export const addChildToSender = (parent: string, child: string): string =>
  joinSenders([parent, child]);
