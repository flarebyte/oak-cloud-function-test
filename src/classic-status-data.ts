import { OakStatus } from './model';

const status_flags = {
  success: 'success',
  failure: 'failure',
  invalid: 'invalid'
}

const ok: OakStatus = {
  name: 'ok',
  description: 'The request has succeeded',
  flags: [status_flags.success],
};

const created: OakStatus = {
  name: 'created',
  description: 'The request has been fulfilled and has resulted in one or more new resources being created',
  flags: [status_flags.success],
};

const accepted: OakStatus = {
  name: 'accepted',
  description: 'The request has been accepted for processing but has not yet completed',
  flags: [status_flags.success],
};

const notFound: OakStatus = {
  name: 'not-found',
  description: 'The document was not found',
  flags: [status_flags.invalid],
};
const badRequest: OakStatus = {
  name: 'bad-request',
  description: 'Malformed request syntax',
  flags: [status_flags.invalid],
};

const unauthorized: OakStatus = {
  name: 'unauthorized',
  description: 'Invalid authentication credentials',
  flags: [status_flags.invalid],
};

const forbidden: OakStatus = {
  name: 'forbidden',
  description: 'Insufficient access rights',
  flags: [status_flags.invalid],
};

const payloadTooLarge: OakStatus = {
  name: 'payload-too-large',
  description: 'The request payload is too large',
  flags: [status_flags.invalid],
};

const unsupportedMediaType: OakStatus = {
  name: 'unsupported-media-type',
  description: 'The payload is not in the right format',
  flags: [status_flags.invalid],
};

const internalServiceError: OakStatus = {
  name: 'internal-service-error',
  description: 'Encountered an unexpected condition',
  flags: [status_flags.failure],
};

const serviceUnavailable: OakStatus = {
  name: 'service-unavailable',
  description: 'Unable to handle the request due to a temporary overload or scheduled maintenance',
  flags: [status_flags.failure],
};

const classicStatusDict = {
  ok,
  created,
  accepted,
  notFound,
  badRequest,
  unauthorized,
  forbidden,
  payloadTooLarge,
  unsupportedMediaType,
  internalServiceError,
  serviceUnavailable

}

const classicStatusList: OakStatus[] = Object.values(classicStatusDict);

export { classicStatusDict, classicStatusList };
