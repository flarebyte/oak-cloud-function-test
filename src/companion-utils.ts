import {
  OakAction,
  OakActionCall,
  OakActionCallWrapper,
  OakActionCompanion,
  OakActionCtxCall,
  OakActionRequestEvent,
  OakActionToCall,
  OakCall,
  OakCallWrapper,
  OakCtxCall,
  OakEngineContext,
  OakFunctionCompanion,
  OakRequestEvent,
  OakServiceOperation,
  OakServiceOpToCall,
} from './model';

const mergeTwoActionCompanion = (
  a: OakActionCompanion,
  b: OakActionCompanion
): OakActionCompanion => ({
  explicitCall: { ...a.explicitCall, ...b.explicitCall },
  call: { ...a.call, ...b.call },
  callServiceOperationDict: {
    ...a.callServiceOperationDict,
    ...b.callServiceOperationDict,
  },
});

export const mergeActionCompanions = (
  companions: OakActionCompanion[]
): OakActionCompanion => companions.reduce(mergeTwoActionCompanion);

type OakServiceOperationDict = {
  [name: string]: OakServiceOperation;
};

const wrapCall =
  (serviceOpDict: OakServiceOperationDict, wrapper: OakCallWrapper) =>
  (keyCall: [string, OakCall]): [string, OakCtxCall] =>
    [keyCall[0], wrapper(serviceOpDict[keyCall[0]], keyCall[1])];

export const transformActionCompanion =
  (wrapper: OakCallWrapper) =>
  (companion: OakActionCompanion): OakActionCompanion => {
    const explicitCall = companion.explicitCall;
    const call = Object.fromEntries(
      Object.entries(explicitCall).map(
        wrapCall(companion.callServiceOperationDict, wrapper)
      )
    );

    return {
      explicitCall,
      call,
      callServiceOperationDict: companion.callServiceOperationDict,
    };
  };

export const buildActionCompanion = (
  mapping: OakServiceOpToCall[]
): OakActionCompanion => {
  const explicitCall = Object.fromEntries(
    mapping.map((m) => [m.so.functionName, m.call])
  );
  const tempCtx: OakEngineContext = {
    transactions: [],
    actionTransactions: [],
    systemFlags: [],
    businessOperationFlags: {},
    actionFlags: {},
    functionFlags: {},
  };
  const call = Object.fromEntries(
    mapping.map((m) => [
      m.so.functionName,
      (req: OakRequestEvent) => m.call(tempCtx, req),
    ])
  );
  const callServiceOperationDict = Object.fromEntries(
    mapping.map((m) => [m.so.functionName, m.so])
  );
  return {
    explicitCall,
    call,
    callServiceOperationDict,
  };
};

const mergeTwoFunctionCompanion = (
  a: OakFunctionCompanion,
  b: OakFunctionCompanion
): OakFunctionCompanion => ({
  explicitCall: { ...a.explicitCall, ...b.explicitCall },
  call: { ...a.call, ...b.call },
  actionDict: { ...a.actionDict, ...b.actionDict },
});

export const mergeFunctionCompanions = (
  companions: OakFunctionCompanion[]
): OakFunctionCompanion => companions.reduce(mergeTwoFunctionCompanion);

type OakActionDict = {
  [name: string]: OakAction;
};

const wrapActionCall =
  (actionDict: OakActionDict, wrapper: OakActionCallWrapper) =>
  (keyCall: [string, OakActionCall]): [string, OakActionCtxCall] =>
    [keyCall[0], wrapper(actionDict[keyCall[0]], keyCall[1])];

export const transformFunctionCompanion =
  (wrapper: OakActionCallWrapper) =>
  (companion: OakFunctionCompanion): OakFunctionCompanion => {
    const explicitCall = companion.explicitCall;
    const call = Object.fromEntries(
      Object.entries(explicitCall).map(
        wrapActionCall(companion.actionDict, wrapper)
      )
    );

    return {
      explicitCall,
      call,
      actionDict: companion.actionDict,
    };
  };

export const buildFunctionCompanion =
  (mapping: OakActionToCall[]) =>
  (actionCompanion: OakActionCompanion): OakFunctionCompanion => {
    const explicitCall = Object.fromEntries(
      mapping.map((m) => [m.action.functionName, m.call])
    );
    const tempCtx: OakEngineContext = {
      transactions: [],
      actionTransactions: [],
      systemFlags: [],
      businessOperationFlags: {},
      actionFlags: {},
      functionFlags: {},
    };
    const call = Object.fromEntries(
      mapping.map((m) => [
        m.action.functionName,
        (req: OakActionRequestEvent) => m.call(tempCtx, actionCompanion, req),
      ])
    );
    const actionDict = Object.fromEntries(
      mapping.map((m) => [m.action.functionName, m.action])
    );
    return {
      explicitCall,
      call,
      actionDict,
    };
  };

export type OakFunctionCompanionBuilder = (
  companion: OakActionCompanion
) => OakFunctionCompanion;
