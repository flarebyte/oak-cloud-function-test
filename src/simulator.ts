import {
  OakEventTransaction,
  OakEventTransactionFilter,
  OakRequestEvent,
  OakResponseEvent,
  OakServiceOperation,
  OakSimulatedCall,
} from './model';
import { summarizeServiceOpTransaction } from './simulator-summarizer';

function cloneValue<A>(value: A): A {
  const jsonstr = JSON.stringify(value);
  const result: A = JSON.parse(jsonstr);
  return result;
}
interface OakSimulatedServiceOpToCall {
  serviceOperation: OakServiceOperation;
  simulatedCall: OakSimulatedCall;
}

class OakSimulator {
  eventTransactions: OakEventTransaction[];
  serviceOpToCall: OakSimulatedServiceOpToCall[];

  constructor() {
    this.eventTransactions = [];
    this.serviceOpToCall = [];
  }
  reset() {
    this.eventTransactions = [];
  }

  addTransaction(request: OakRequestEvent, response: OakResponseEvent) {
    const transaction = {
      id: this.eventTransactions.length,
      request,
      response,
    };
    this.eventTransactions.push(cloneValue(transaction));
  }

  registerServiceOpToCall(
    serviceOperation: OakServiceOperation,
    simulatedCall: OakSimulatedCall
  ) {
    this.serviceOpToCall.push({ serviceOperation, simulatedCall });
  }

  toFullInfo() {
    return JSON.stringify(this.eventTransactions, null, 2);
  }

  toInfo() {
    return JSON.stringify(
      this.eventTransactions.map(summarizeServiceOpTransaction),
      null,
      2
    );
  }

  serviceCall(serviceOperation: OakServiceOperation, request: OakRequestEvent) {
    const mapping = this.serviceOpToCall.find(
      m => m.serviceOperation.name === serviceOperation.name
    );
    if (!mapping)
      throw Error(`No hook for service operation:  ${serviceOperation.name}`);
    const result = mapping.simulatedCall(this.eventTransactions, request);
    return result;
  }

  filterTransactions(
    tFilter: OakEventTransactionFilter
  ): OakEventTransaction[] {
    return this.eventTransactions.filter(tFilter);
  }
}

const simulatedCall = (simulator: OakSimulator) => async (
  reqEvent: OakRequestEvent
) => {
  const respEvent = simulator.serviceCall(
    reqEvent.businessOperation.serviceOperation,
    reqEvent
  );
  simulator.addTransaction(reqEvent, respEvent);
  return Promise.resolve(respEvent);
};

export { simulatedCall, OakSimulator };
