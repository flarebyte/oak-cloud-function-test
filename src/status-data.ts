import { OakStatus } from './model';

export const statusFlagDict = {
  success: 'success',
  failure: 'failure',
  invalid: 'invalid',
};

const ok: OakStatus = {
  name: 'ok',
  description: 'The request has succeeded',
  flags: [statusFlagDict.success],
};

const found: OakStatus = {
  name: 'found',
  description: 'The resource has been found',
  flags: [statusFlagDict.success],
};

const absent: OakStatus = {
  name: 'absent',
  description: 'The resource is not present',
  flags: [statusFlagDict.success],
};

const empty: OakStatus = {
  name: 'empty',
  description: 'The collection has been found but is empty',
  flags: [statusFlagDict.success],
};

const single: OakStatus = {
  name: 'single',
  description: 'A collection has been found with a single element',
  flags: [statusFlagDict.success],
};

const multiple: OakStatus = {
  name: 'multiple',
  description: 'A collection has been found with a multiple elements',
  flags: [statusFlagDict.success],
};

const created: OakStatus = {
  name: 'created',
  description:
    'The request resulted in the creation of one or more new resources',
  flags: [statusFlagDict.success],
};

const updated: OakStatus = {
  name: 'updated',
  description:
    'The request resulted in the update of one or more new resources',
  flags: [statusFlagDict.success],
};

const createdOrUpdated: OakStatus = {
  name: 'created-or-updated',
  description:
    'The request resulted in the creation or update of one or more resources',
  flags: [statusFlagDict.success],
};

const deleted: OakStatus = {
  name: 'deleted',
  description: 'The request resulted in the deletion of one or more resources',
  flags: [statusFlagDict.success],
};

const accepted: OakStatus = {
  name: 'accepted',
  description:
    'The request has been accepted for processing but has not yet completed',
  flags: [statusFlagDict.success],
};

const notFound: OakStatus = {
  name: 'not-found',
  description: 'The document was not found but was expected',
  flags: [statusFlagDict.invalid],
};

const badRequest: OakStatus = {
  name: 'bad-request',
  description: 'Malformed request syntax',
  flags: [statusFlagDict.invalid],
};

const unauthorized: OakStatus = {
  name: 'unauthorized',
  description: 'Invalid authentication credentials',
  flags: [statusFlagDict.invalid],
};

const forbidden: OakStatus = {
  name: 'forbidden',
  description: 'Insufficient access rights',
  flags: [statusFlagDict.invalid],
};

const payloadTooLarge: OakStatus = {
  name: 'payload-too-large',
  description: 'The request payload is too large',
  flags: [statusFlagDict.invalid],
};

const unsupportedMediaType: OakStatus = {
  name: 'unsupported-media-type',
  description: 'The payload is not in the right format',
  flags: [statusFlagDict.invalid],
};

const internalServiceError: OakStatus = {
  name: 'internal-service-error',
  description: 'Encountered an unexpected condition',
  flags: [statusFlagDict.failure],
};

const serviceUnavailable: OakStatus = {
  name: 'service-unavailable',
  description:
    'Unable to handle the request due to a temporary overload or scheduled maintenance',
  flags: [statusFlagDict.failure],
};

export const statusDict = {
  ok,
  found,
  absent,
  empty,
  single,
  multiple,
  created,
  updated,
  createdOrUpdated,
  accepted,
  deleted,
  notFound,
  badRequest,
  unauthorized,
  forbidden,
  payloadTooLarge,
  unsupportedMediaType,
  internalServiceError,
  serviceUnavailable,
};
