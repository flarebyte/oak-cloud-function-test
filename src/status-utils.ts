import { OakStatus } from './model';
import { statusFlagDict } from './status-data';

export const isStatusEqual = (
  expected: OakStatus,
  actual: OakStatus
): boolean => actual.name === expected.name;

export const isSuccessfulStatus = (actual: OakStatus): boolean =>
  actual.flags.includes(statusFlagDict.success);

export const isInvalidStatus = (actual: OakStatus): boolean =>
  actual.flags.includes(statusFlagDict.invalid);

export const isFailureStatus = (actual: OakStatus): boolean =>
  actual.flags.includes(statusFlagDict.failure);

export const findStatusByName = (
  name: string,
  statusList: OakStatus[]
): OakStatus | false =>
  statusList.find((status) => status.name === name) || false;

export const findStatusByNameInDict = (
  name: string,
  refStatusDict: { [key: string]: OakStatus }
): OakStatus | false => findStatusByName(name, Object.values(refStatusDict));
