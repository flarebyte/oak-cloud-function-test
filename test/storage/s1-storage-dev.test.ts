import { OakRequestEvent } from '../../src/model';
import { bizOperationObj } from '../fixture-business-operation';
import { s1DevCompanion } from './s1-storage-dev';
import { createS1Params } from './s1-storage-factory';
import { OakSimulator } from '../../src/simulator';
import { isSuccessfulStatus } from '../../src/status-utils';

import { statusDict } from '../../src/status-data';

const { notFound, badRequest, forbidden } = statusDict;

const writeToLondonRequestTemplate = (
  path: string,
  value: number
): OakRequestEvent => ({
  businessOperation: bizOperationObj.writeLondonData,
  caller: 'test',
  comment: 'write to london',
  serviceParams: createS1Params(path),
  systemFlags: [],
  payload: {
    data: [1, 2, 3, value],
  },
  flags: [],
});

const writeToParisRequestTemplate = (
  path: string,
  value: number
): OakRequestEvent => ({
  businessOperation: bizOperationObj.writeParisData,
  caller: 'test',
  comment: 'write to Paris',
  serviceParams: createS1Params(path),
  systemFlags: [],
  payload: {
    data: [1, 2, 3, value],
  },
  flags: [],
});

const readFromLondonRequestTemplate = (path: string): OakRequestEvent => ({
  businessOperation: bizOperationObj.readLondonData,
  caller: 'test',
  comment: 'read to london',
  serviceParams: createS1Params(path),
  systemFlags: [],
  payload: {},
  flags: [],
});

describe('S1 Storage', () => {
  const simulator = new OakSimulator();
  simulator.registerCompanions([s1DevCompanion], []);
  const call = simulator.getCall();

  beforeEach(() => {
    simulator.reset();
  });
  describe('Test write', () => {
    it('should support multiple write', async () => {
      const resp1 = await call.writeS1(
        writeToLondonRequestTemplate('london/city/data', 12)
      );

      const resp2 = await call.writeS1(
        writeToLondonRequestTemplate('london/city/data', 12)
      );
      expect(isSuccessfulStatus(resp1.status)).toBeTruthy();
      expect(isSuccessfulStatus(resp2.status)).toBeTruthy();
    });

    it('should return bad request if parameters are incorrect', async () => {
      const resp = await call.writeS1({
        ...writeToLondonRequestTemplate('london/city/data', 10),
        serviceParams: { unknownParam: 'some value' },
      });
      expect(resp.status.name).toEqual(badRequest.name);
    });

    it('should return forbidden if forbidden', async () => {
      simulator.setReturnStatusForBusinessOp(
        bizOperationObj.writeLondonData,
        forbidden
      );
      const resp = await call.writeS1({
        ...writeToLondonRequestTemplate('london/city/data', 11),
      });
      expect(resp.status.name).toEqual(forbidden.name);
    });
  });

  describe('Test read', () => {
    it('should read the latest value for path', async () => {
      await call.writeS1(writeToLondonRequestTemplate('london/city/data', 17));

      await call.writeS1(writeToLondonRequestTemplate('london/city/data', 19));
      const respRead = await call.readS1(
        readFromLondonRequestTemplate('london/city/data')
      );
      expect(isSuccessfulStatus(respRead.status)).toBeTruthy();
      expect(respRead.payload).toHaveProperty('data', [1, 2, 3, 19]);
      expect(simulator.toInfo()).toMatchSnapshot();
    });
    it('should return not found if the value is not present', async () => {
      const resp = await call.readS1(
        readFromLondonRequestTemplate('oxford/city/data')
      );
      expect(resp.status.name).toEqual(notFound.name);
    });
    it('should read the value from the right bucket', async () => {
      await call.writeS1(writeToParisRequestTemplate('london/city/data', 17));

      const respRead = await call.readS1(
        readFromLondonRequestTemplate('london/city/data')
      );
      expect(respRead.status.name).toEqual(notFound.name);
    });
  });
});
