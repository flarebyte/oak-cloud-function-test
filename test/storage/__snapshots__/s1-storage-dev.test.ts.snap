// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S1 Storage Test read should read the latest value for path 1`] = `
"[
  {
    \\"id\\": 0,
    \\"businessOperationName\\": \\"write-to-london\\",
    \\"serviceOperationName\\": \\"company:s1:write\\",
    \\"serviceName\\": \\"company:s1\\",
    \\"resourceName\\": \\"s3:london\\",
    \\"caller\\": \\"test\\",
    \\"serviceParams\\": {
      \\"path\\": \\"london/city/data\\"
    },
    \\"systemFlags\\": \\"\\",
    \\"reqComment\\": \\"write to london\\",
    \\"reqPayload\\": {
      \\"data\\": [
        1,
        2,
        3,
        17
      ]
    },
    \\"reqFlags\\": \\"\\",
    \\"statusName\\": \\"ok\\",
    \\"respComment\\": \\"Success\\",
    \\"respPayload\\": {
      \\"message\\": \\"Saved\\"
    },
    \\"respFlags\\": \\"\\"
  },
  {
    \\"id\\": 1,
    \\"businessOperationName\\": \\"write-to-london\\",
    \\"serviceOperationName\\": \\"company:s1:write\\",
    \\"serviceName\\": \\"company:s1\\",
    \\"resourceName\\": \\"s3:london\\",
    \\"caller\\": \\"test\\",
    \\"serviceParams\\": {
      \\"path\\": \\"london/city/data\\"
    },
    \\"systemFlags\\": \\"\\",
    \\"reqComment\\": \\"write to london\\",
    \\"reqPayload\\": {
      \\"data\\": [
        1,
        2,
        3,
        19
      ]
    },
    \\"reqFlags\\": \\"\\",
    \\"statusName\\": \\"ok\\",
    \\"respComment\\": \\"Success\\",
    \\"respPayload\\": {
      \\"message\\": \\"Saved\\"
    },
    \\"respFlags\\": \\"\\"
  },
  {
    \\"id\\": 2,
    \\"businessOperationName\\": \\"read-from-london\\",
    \\"serviceOperationName\\": \\"company:s1:read\\",
    \\"serviceName\\": \\"company:s1\\",
    \\"resourceName\\": \\"s3:london\\",
    \\"caller\\": \\"test\\",
    \\"serviceParams\\": {
      \\"path\\": \\"london/city/data\\"
    },
    \\"systemFlags\\": \\"\\",
    \\"reqComment\\": \\"read to london\\",
    \\"reqPayload\\": {},
    \\"reqFlags\\": \\"\\",
    \\"statusName\\": \\"ok\\",
    \\"respComment\\": \\"Success\\",
    \\"respPayload\\": {
      \\"data\\": [
        1,
        2,
        3,
        19
      ]
    },
    \\"respFlags\\": \\"\\"
  }
]"
`;
